[tools] fix Android Studio duplicate detection when installed via Jetbrains Toolbox

TahaTesser opened the pull request:
Added condition to check if detected `Android Studio.app` has correct executable and tests to avoid duplicate Android Studio detections when installed via Jetbrains 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

TahaTesser committed:
fix Android Studio duplicates

TahaTesser committed:
remove whitespace

jmagman commented on packages/flutter_tools/lib/src/android/android_studio.dart:
```diff
@@ -475,4 +476,14 @@ class AndroidStudio implements Comparable<AndroidStudio> {
+  @override
+  // ignore: avoid_equals_and_hash_code_on_mutable_classes
+  bool operator ==(Object other) {
+    return other is Object && hashCode == other.hashCode;
```
`==` should check the other object is also an `AndroidStudio` object, not `Object`.  `==` needs to actually check equality, not hashCodes.  `==` needs to be reliable even with hash collisions.
```suggestion
    return other is AndroidStudio && directory == other.directory;
```


> // ignore: avoid_equals_and_hash_code_on_mutable_classes

I believe this class could be made immutable if the `_validationMessages` population were hoisted into the factory method, then `_validationMessages`, and `_isValid`  passed in as an immutable `List`.  However, instead of doing all that, I think the right thing is to remove your `hashCode` and `==` changes, and instead fix the existing `FileSystemEntity` checks.

jmagman reviewed (changes requested):
1. As the auto generated comment says, this needs tests.  Check out https://github.com/flutter/flutter/blob/94f71f374f904de336d81fcdcce3a48cbf26639a/packages/flutter_tools/test/general.shard/android/android_studio_test.dart#L180
2. Since you're reproducing, can you step through and see why exactly is this happening?  It looks like there are already safeguards in place, so it probably makes sense to fix the bug here
https://github.com/flutter/flutter/blob/94f71f374f904de336d81fcdcce3a48cbf26639a/packages/flutter_tools/lib/src/android/android_studio.dart#L294-L295
and
https://github.com/flutter/flutter/blob/94f71f374f904de336d81fcdcce3a48cbf26639a/packages/flutter_tools/lib/src/android/android_studio.dart#L314-L315
If you step through you should be able to see why this is happening.

TahaTesser committed:
Merge branch 'master' of github.com:flutter/flutter into fix_duplicate_as

TahaTesser committed:
updated fix and added test

TahaTesser commented:
@jmagman 
Updated the PR and added test, the root of the cause is that, there are two `Android Studio.app`, one from the launcher and one from apps folder, to avoid launcher app, I have added condition that checks for expected executable string and added test that simulates this exact issue and test passes with one Android Studio being detected (rather than two)

TahaTesser committed:
fix condition nullalibity

jmagman commented:
> one from the launcher and one from apps folder, to avoid launcher app

Is one a symlink?

TahaTesser commented:
@jmagman 
Path for launcher app is `/Users/tahatesser/Applications/JetBrains Toolbox/Android Studio.app`, I am not sure this is a symlink. This app only has launcher resources such as an icon and some metadata that points to  actual location but in flutter_tools, we're treating it as a real Android Studio app since it ends with the same name but this PR avoids that by checking  if the app's info list has an executable `studio`  

<img width="1233" alt="Screen Shot 2021-08-09 at 20 40 08" src="https://user-images.githubusercontent.com/48603081/128750169-71eecf6b-7ab4-44c1-98dd-be791e3cfc7c.png">

TahaTesser commented:
Launcher app is only used here and has a size less than 500kb

<img width="471" alt="Screen Shot 2021-08-09 at 20 50 43" src="https://user-images.githubusercontent.com/48603081/128751055-c536ce4e-30cf-4d6a-a5a0-c9cc5f79ad86.png">

jmagman reviewed (commented):
> This app only has launcher resources such as an icon and some metadata that points to actual location but in flutter_tools, we're treating it as a real Android Studio app since it ends with the same name but this PR avoids that by checking if the app's info list has an executable `studio`

I understand avoiding the placeholder launcher bundle.  But why is it duplicated though?

TahaTesser committed:
restore empty line

TahaTesser commented:
> > This app only has launcher resources such as an icon and some metadata that points to actual location but in flutter_tools, we're treating it as a real Android Studio app since it ends with the same name but this PR avoids that by checking if the app's info list has an executable `studio`
> 
> I understand avoiding the placeholder launcher bundle. But why is it duplicated though?

One of them is detected from directories scan, which I fixed and the other one detected from mfind so the array size is increased to two and we get duplicates as a result.

If we comment out directories scan, we only get one detection

TahaTesser committed:
restore 2

TahaTesser committed:
restore 3

TahaTesser committed:
restore

TahaTesser commented on packages/flutter_tools/test/general.shard/android/android_studio_test.dart:
```diff
@@ -157,6 +174,7 @@ void main() {
       Platform: () => platform,
       PlistParser: () => plistUtils,
     });
+
```
@jmagman 
In my other PR, I forgot to add extra line here so fixed it here

jmagman commented:
> One of them is detected from directories scan, which I fixed and the other one detected from mfind so the array size is increased to two and we get duplicates as a result.

I don't think the solution is to peg to a particular executable name since we don't _really_ care about that, and this would break if this ever changed.  What is `candidatePaths` is made into a `Set`?

TahaTesser commented:
I agree, it looks unnecessary, I have toSet and `==`, it does solve the issue however I am not sure what proper changes are needed to make the class immutable so I will close the PR, Appreciate your review, learned a lot from this PR

jmagman commented:
> I agree, it looks unnecessary, I have toSet and `==`, it does solve the issue however I am not sure what proper changes are needed to make the class immutable so I will close the PR, Appreciate your review, learned a lot from this PR

I think you were close!  I want to make sure we are fixing the right thing though.  `candidatePaths` is currently a `List<FileSystemEntity>` so it doesn't need anything to be immutable.

TahaTesser commented:
@jmagman 
Applying to toSet to `List<FileSystemEntity>` doesn't have an impact since both locations are already unique in List<FileSystemEntity> (one from the toolbox folder and one from apps)
 I confirmed it by printing it before we map this list of `AndroidStudio`
However when applied toSet to `AndroidStudio` and add `==`, issues are resolved

I want to contribute more and better, so I am going to study the inner workings for a few weeks and then send better PR. I also wanna fix VS Code 2022 issue but not before I completely understand the file system, classes, etc :)

jmagman commented:
> Applying to toSet to `List<FileSystemEntity>` doesn't have an impact since both locations are already unique in List (one from the toolbox folder and one from apps)

I get it now.  https://github.com/flutter/flutter/issues/86291#issuecomment-899708390
> I see the problem. The bundle finder is discovering `~/Applications/JetBrains Toolbox/Android Studio.app` and `~/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/203.7583922/Android Studio.app`. The former has its Info.plist `JetBrainsToolboxApp` is pointing to the latter `~/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/203.7583922/Android Studio.app`. The latter does not have a `JetBrainsToolboxApp` in its Info.plist
> So the path discovery logic works as of the addition of #80475, but the `AndroidStudio.fromMacOSBundle(e.path)` is finding both.


> I want to contribute more and better, so I am going to study the inner workings for a few weeks and then send better PR.

Awesome!  Instead of dealing with sets, I think we can instead get rid of the logic that parses that Info.plist file to follow the `JetBrainsToolboxApp` path since as of #80475 it already handles finding the "real" Android Studio via Spotlight.

A way to do this would be to change https://github.com/flutter/flutter/blob/f9aa5842c285686fc553976e68d60f4351963191/packages/flutter_tools/lib/src/android/android_studio.dart#L319-L322 to:
```dart
    return candidatePaths
        .map<AndroidStudio?>((FileSystemEntity e) => AndroidStudio.fromMacOSBundle(e.path))
        .whereType<AndroidStudio>()
        .toList();
```
and https://github.com/flutter/flutter/blob/f9aa5842c285686fc553976e68d60f4351963191/packages/flutter_tools/lib/src/android/android_studio.dart#L43-L54 to
```dart
  static AndroidStudio? fromMacOSBundle(String bundlePath) {
    final String studioPath = globals.fs.path.join(bundlePath, 'Contents');
    final String plistFile = globals.fs.path.join(studioPath, 'Info.plist');
    final Map<String, dynamic> plistValues = globals.plistParser.parseFile(plistFile);
    // As AndroidStudio managed by JetBrainsToolbox could have a wrapper pointing to the real Android Studio.
    // If we've found a JetBrainsToolbox wrapper, ignore it.
    if (plistValues.containsKey('JetBrainsToolboxApp')) {
      return null;
    }
```
Writing tests for this left as an exercise to the reader.  😄 

Let me know if you need any help!

TahaTesser commented:
@jmagman 
I would like to work on it, file a new PR by tomorrow. I really appreciate your time
